File: lee/src/main/java/org/example/lee/题目/二分查找/b_搜索旋转排序数组.java
package org.example.lee.题目.二分查找;

import java.util.Arrays;

public class b_搜索旋转排序数组 {


  /**
   * search 要求 logN 二分法 然后找出有序的空间和无序的空间 有序则继续二分 无序则递归
   *
   * @param nums   nums
   * @param target target
   * @return int
   */
  public static int search(int[] nums, int target) {
    int len = nums.length, mid, left = 0, right = len - 1, res1 = -1, res2 = -1;
    mid = (left + right) / 2;
    if (nums[mid] == target) {
      return mid;
    } else {
      if (nums[mid] > nums[right]) {
        res1 = search(Arrays.copyOfRange(nums, mid + 1, len), target);
        res2 = tool(nums, target, 0, mid - 1);
        if (res1 != -1) {
          return mid + 1 + res1;
        }
        return res2;
      } else if (nums[mid] < nums[left]) {
        res1 = search(Arrays.copyOfRange(nums, 0, mid), target);
        res2 = tool(nums, target, mid + 1, len - 1);
        if (res1 != -1) {
          return res1;
        }
        return res2;
      } else {
        return tool(nums, target, 0, len - 1);
      }

    }
  }

  public static int tool(int[] nums, int target, int left, int right) {
    int middle = 0;
    do {
      middle = (left + right) / 2;
      if (nums[middle] > target) {
        right = middle - 1;
      } else if (nums[middle] < target) {
        left = middle + 1;
      } else {
        return middle;
      }
    } while (left <= right);
    return -1;
  }
}

File: lee/src/main/java/org/example/lee/题目/二分查找/b_搜索旋转排序数组.go
package 二分查找

func search(nums []int, target int) int {
	return search1(nums, 0, len(nums)-1, target)
}

func search1(nums []int, left, right int, target int) int {
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] == target {
			return mid
		} else {
			//左边数组有序
			if nums[left] <= nums[mid] {
				//target位于左边数组
				if nums[left] <= target && nums[mid] > target {
					right = mid - 1
				} else {
					//target位于右边数组
					left = mid + 1
				}
			} else {
				//右边数组有序
				//target位于右边数组
				if nums[right] >= target && nums[mid] < target {
					left = mid + 1
				} else {
					//target位于左边数组
					right = mid - 1
				}
			}
		}
	}
	return -1
}

//TODO 垃圾版本
//func search0(nums []int, left, right, target int) int {
//	if left > right {
//		return -1
//	}
//	mid := left + (right-left)/2
//	if nums[mid] == target {
//		return mid
//	} else if nums[mid] > nums[right] {
//		i := search0(nums, mid+1, right, target)
//		if i != -1 {
//			return i
//		}
//	} else if nums[mid] < nums[left] {
//		i := search0(nums, left, mid-1, target)
//		if i != -1 {
//			return i
//		}
//	}
//	return midSearch(nums, left, right, target)
//}
//func midSearch(nums []int, left, right, target int) int {
//	for left <= right {
//		mid := left + (right-left)/2
//		if nums[mid] == target {
//			return mid
//		} else if nums[mid] < target {
//			left = mid + 1
//		} else {
//			right = mid - 1
//		}
//	}
//	return -1
//}
